src/lifetimes.rs

```diff
@@ -0,0 +1,134 @@
+use syntax::ast::*;
+use rustc::lint::{Context, LintPass, LintArray, Lint};
+use syntax::codemap::Span;
+use syntax::visit::{Visitor, FnKind, walk_ty};
+use utils::{in_external_macro, span_lint};
+use std::collections::HashSet;
+use std::iter::FromIterator;
+
+declare_lint!(pub NEEDLESS_LIFETIMES, Warn,
+              "Warn on explicit lifetimes when elision rules would apply");
+
+#[derive(Copy,Clone)]
+pub struct LifetimePass;
+
+impl LintPass for LifetimePass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(NEEDLESS_LIFETIMES)
+    }
+
+    fn check_fn(&mut self, cx: &Context, kind: FnKind, decl: &FnDecl,
+                _: &Block, span: Span, _: NodeId) {
+        if in_external_macro(cx, span) {
+            return;
+        }
+        if could_use_elision(kind, decl) {
+            span_lint(cx, NEEDLESS_LIFETIMES, span,
+                      "explicit lifetimes given in parameter types where they could be elided");
+        }
+    }
+}
+
+/// The lifetime of a &-reference.
+#[derive(PartialEq, Eq, Hash, Debug)]
+enum RefLt {
+    Unnamed,
+    Static,
+    Named(Name),
+}
+use self::RefLt::*;
+
+fn could_use_elision(kind: FnKind, func: &FnDecl) -> bool {
+    // There are two scenarios where elision works:
+    // * no output references, all input references have different LT
+    // * output references, exactly one input reference with same LT
+
+    // these will collect all the lifetimes for references in arg/return types
+    let mut input_visitor = RefVisitor(Vec::new());
+    let mut output_visitor = RefVisitor(Vec::new());
+
+    // extract lifetime in "self" argument for methods (there is a "self" argument
+    // in func.inputs, but its type is TyInfer)
+    if let FnKind::FkMethod(_, sig, _) = kind {
+        match sig.explicit_self.node {
+            SelfRegion(ref opt_lt, _, _) => input_visitor.record(opt_lt),
```

Comment:
We should additionally walk the inner type, e.g. `&'a Foo<'b>` and `&'a Foo<'a>` have different effects, and we should only be suggesting elision for the former